### PR TITLE
Refactor Elasticsearch Integration

### LINKED ISSUE #522: Search Index Handling & Synchronization
# Abstract

Index operations (Create, Update, Delete) should happen async. Sync operations currently impact write request performance. Sync operations are currently required since Gentics Mesh does not yet support Rest filtering #485.

# Tasks

* Replace sync with async operations
* Add circuit breakers to detect index operation failures
* Index operations failures should trigger a recovery mode in which a sync should be executed
* Add verticle for recovery operations. The verticle should periodically check whether the needed indices exist. Unknown indices with the configured prefix should be removed.
* Gentics Mesh should keep a reference of the current index state to automatically pick up the sync process. This should be done to avoid full-sync operations. The sync state needs to be stored in a dedicated ES index or in the index metadata if possible.

# Operation Handling

* Use regular mesh events to determine which operations need to be executed on the elasticsearch indices.
* Add root cause to events. We can use the root cause to skip certain events. (e.g. Project deletion does not need to remove documents from the index. Instead the whole index can be dropped)
* Operations need to be batches for a given time window. 
* Batch processing will be triggered once the batch size limit has been reached.
* Batch processing can also be triggered once a schema migration completed event has been received.
* The tx handlers of mesh rest operations collects the events (similar to SQB's). This collection will be processed once the tx has been comitted in order to trigger the events.

### PR DESCRIPTION
**Pull request by Jotschi**
Resolves #522

### CODE REVIEW

**Comment by Jotschi** [comment]
@philippguertler 
Very important: We need to add 429 error handling to ES index and sync feature.

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/role/PermissionChangedEventModelImpl.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package com.gentics.mesh.core.rest.event.role;
+
+import com.gentics.mesh.ElementType;
+import com.gentics.mesh.core.rest.event.AbstractElementMeshEventModel;
+import com.gentics.mesh.core.rest.role.RoleReference;
+
+public class PermissionChangedEventModelImpl extends AbstractElementMeshEventModel implements PermissionChangedEventModel {
+
+	private ElementType type;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/role/PermissionChangedEventModelImpl.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/role/PermissionChangedProjectElementEventModel.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.gentics.mesh.core.rest.event.role;
+
+import com.gentics.mesh.core.rest.event.MeshProjectElementEventModel;
+import com.gentics.mesh.core.rest.project.ProjectReference;
+
+public class PermissionChangedProjectElementEventModel extends PermissionChangedEventModelImpl implements MeshProjectElementEventModel {
+
+	private ProjectReference project;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/role/PermissionChangedProjectElementEventModel.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/role/TagPermissionChangedEventModel.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package com.gentics.mesh.core.rest.event.role;
+
+import com.gentics.mesh.core.rest.event.tag.TagElementEventModel;
+import com.gentics.mesh.core.rest.tag.TagFamilyReference;
+
+public class TagPermissionChangedEventModel extends PermissionChangedProjectElementEventModel implements TagElementEventModel {
+
+	private TagFamilyReference tagFamily;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/role/TagPermissionChangedEventModel.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/tag/TagMeshEventModel.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.gentics.mesh.core.rest.event.tag;
+
+import com.gentics.mesh.core.rest.event.AbstractProjectEventModel;
+import com.gentics.mesh.core.rest.tag.TagFamilyReference;
+
+public class TagMeshEventModel extends AbstractProjectEventModel implements TagElementEventModel {
+
+	private TagFamilyReference tagFamily;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/tag/TagMeshEventModel.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/group/GroupUserAssignModel.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package com.gentics.mesh.core.rest.event.group;
+
+import com.gentics.mesh.core.rest.event.AbstractMeshEventModel;
+import com.gentics.mesh.core.rest.group.GroupReference;
+import com.gentics.mesh.core.rest.user.UserReference;
+
+public class GroupUserAssignModel extends AbstractMeshEventModel {
+
+	private GroupReference group;
+	private UserReference user;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/group/GroupUserAssignModel.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/group/GroupRoleAssignModel.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.gentics.mesh.core.rest.event.group;
+
+import com.gentics.mesh.core.rest.event.AbstractMeshEventModel;
+import com.gentics.mesh.core.rest.group.GroupReference;
+import com.gentics.mesh.core.rest.role.RoleReference;
+
+public class GroupRoleAssignModel extends AbstractMeshEventModel {
+
+	private GroupReference group;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/group/GroupRoleAssignModel.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/EventCauseInfoImpl.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
+package com.gentics.mesh.core.rest.event;
+
+import com.gentics.mesh.ElementType;
+
+public class EventCauseInfoImpl implements EventCauseInfo {
+
+	private ElementType type;
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/EventCauseInfoImpl.java]: attributes

File: rest-model/src/main/java/com/gentics/mesh/core/rest/event/EventCauseAction.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package com.gentics.mesh.core.rest.event;
+
+import com.gentics.mesh.core.rest.event.node.BranchMigrationCause;
+import com.gentics.mesh.core.rest.event.node.MicroschemaMigrationCause;
+import com.gentics.mesh.core.rest.event.node.SchemaMigrationCause;
+
+public enum EventCauseAction {
```
Comment (Jotschi) [inline rest-model/src/main/java/com/gentics/mesh/core/rest/event/EventCauseAction.java]: javadoc

File: core/src/test/resources/logback-test.xml [hunk 9]
```diff
@@ -41,6 +41,8 @@
 	<logger name="com.gentics.mesh.core.binary.impl" level="DEBUG" />
 	 -->
 	<!-- <logger name="com.gentics.mesh.core.data.search.impl.SearchQueueBatchImpl" level="DEBUG" /> -->
+	<logger name="com.gentics.mesh.search.index" level="DEBUG" />
+	<logger name="com.gentics.mesh.search.verticle" level="TRACE" />
```
Comment (Jotschi) [inline core/src/test/resources/logback-test.xml]: Remove trace

**Comment by Jotschi** [comment]
Commit generated includes.

**Comment by philippguertler** [comment]
Add better documentation for the eventbus protocol

**Comment by Jotschi** [comment]
Merged into dev

**Pull request closed without merging**